AX: [AX-Thread Text APIs] OffsetFromRoot should explore all text objects, not just those with text runs

hoffmanjoshua opened the pull request:
#### b0fe6c1717f59766f7143a5222cbc6b5e4a9da6c
<pre>
AX: [AX-Thread Text APIs] OffsetFromRoot should explore all text objects, not just those with text runs
<a href="https://bugs.webkit.org/show_bug.cgi?id=287736">https://bugs.webkit.org/show_bug.cgi?id=287736</a>
<a href="https://rdar.apple.com/144892240">rdar://144892240</a>

Reviewed by Tyler Wilcock.

ATs sometimes need to request the index of a text marker that is not part of an object with text runs.
For example, the start of a text input that has no text.

OffsetFromRoot used to assert in this scenario, since it only traversed text positions that were in text
runs. Now, we explore all objects, switching between navigating within runs and between objects when
necessary.

* LayoutTests/accessibility/ax-thread-text-apis/offset-from-root-outside-text-run-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/offset-from-root-outside-text-run.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::offsetFromRoot const):

Canonical link: <a href="https://commits.webkit.org/290445@main">https://commits.webkit.org/290445@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7928df768892adb4016be915bd089112e9e7619a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90011 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44911 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95011 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40784 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92063 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17826 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69304 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26900 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93012 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7592 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81650 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49651 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7320 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39918 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77655 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37076 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96837 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17199 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78307 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17455 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77492 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19143 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21944 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10393 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17209 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22534 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16950 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20402 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18733 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -574,15 +574,30 @@ unsigned AXTextMarker::offsetFromRoot() const
+
+                // If we come across an object on a new line, we need to increment the offset, since the previous + current
+                // text marker won't share an equivalent visual text position.
+                // However, If we are moving on or off of a line break, don't compare lineIDs. The line break object has
```
Typo here I think, capital I on If 🙂

hoffmanjoshua replied:
Good catch—fixed!

hoffmanjoshua committed:
AX: [AX-Thread Text APIs] OffsetFromRoot should explore all text objects, not just those with text runs
https://bugs.webkit.org/show_bug.cgi?id=287736
rdar://144892240

Reviewed by Tyler Wilcock.

ATs sometimes need to request the index of a text marker that is not part of an object with text runs.
For example, the start of a text input that has no text.

OffsetFromRoot used to assert in this scenario, since it only traversed text positions that were in text
runs. Now, we explore all objects, switching between navigating within runs and between objects when
necessary.

* LayoutTests/accessibility/ax-thread-text-apis/offset-from-root-outside-text-run-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/offset-from-root-outside-text-run.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::offsetFromRoot const):

Canonical link: https://commits.webkit.org/290445@main
